pkg/store/objectstore/flatfs/flatfs.go

```diff
@@ -0,0 +1,500 @@
+// Package flatfs is a Datastore implementation that stores all
+// objects in a two-level directory structure in the local file
+// system, regardless of the hierarchy of the keys.
+package flatfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/storacha/piri/pkg/store/objectstore"
+)
+
+var log = logging.Logger("flatfs")
+
+const extension = ".data"
+
+var (
+	// RetryDelay is a timeout for a backoff on retrying operations
+	// that fail due to transient errors like too many file descriptors open.
+	RetryDelay = time.Millisecond * 200
+	// RetryAttempts is the maximum number of retries that will be attempted
+	// before giving up.
+	RetryAttempts = 6
+)
+
+const (
+	opPut = iota
+	opDelete
+)
+
+var _ objectstore.Store = (*Store)(nil)
+
+var (
+	ErrStoreExists         = errors.New("datastore already exists")
+	ErrStoreDoesNotExist   = errors.New("datastore directory does not exist")
+	ErrShardingFileMissing = fmt.Errorf("%s file not found in datastore", SHARDING_FN)
+	ErrClosed              = errors.New("datastore closed")
+	ErrInvalidKey          = errors.New("key not supported by flatfs")
+)
+
+// Store implements [objectstore.Store].
+// Note this datastore cannot guarantee order of concurrent
+// write operations to the same key. See the explanation in
+// Put().
+type Store struct {
```

Comment:
A `Has` method would be really nice to have, specifically for https://github.com/storacha/piri/issues/323

Reply:
🙏 please can we add if we really need it? I don't think there's a big overhead with using `Get` for this - it's not going to start reading the file until you call `Body()` on the returned object.